### DISCUSSION TITLE
array of objects conversion into memory structure

### DISCUSSION THREAD

**Discussion by pedro-vicente** [nlohmann/json#3355] [Q&A] [answered]
I am reading a JSON file with this structure

```

    {
      "Main": [
        {
          "obj1": "bar"
        },
        {
          "obj2": "foo"
        }
      ]
    }
```

The main object with key known to the user "Main" contains an array of objects with unknown key names

I want to transfer the JSON object o in the following program to a C++ struture?
How could this be done?

```


    #include <nlohmann/json.hpp>
    #include <iostream>
    #include <fstream>
    using json = nlohmann::json;
    
    int main()
    {
      std::ifstream ifs("../test/test_2.json");
      json js = json::parse(ifs);
    
      for (json::iterator it = js.begin(); it != js.end(); ++it) 
      {
        std::cout << it.key() << " :\n";
        std::cout << it.value() << "\n";
      }
    
      if (js.contains("Main"))
      {
        json a = js["Main"];
        for (size_t idx = 0; idx < a.size(); ++idx)
        {
          json o = a.at(idx);
          std::cout << o << "\n";
        }
      }
      return 0;
    }

```

Output:

```

    Main :
    [{"obj1":"bar"},{"obj2":"foo"}]
    {"obj1":"bar"}
    {"obj2":"foo"}
    
```

**Comment by nlohmann** [comment]
I don't understand what you want to achieve.

**Reply by pedro-vicente** [reply]
I want to store the objects in the array (each one with variable json o) in a vector of , for example, this C++ structure


```
struct field_value_t
{
  std::string field;
  std::string value;
}; 
```

**Reply by pedro-vicente** [reply]
The examples in the API show how to get an object with a known name
```

void from_json(const json& j, person& p) {
        j.at("name").get_to(p.name);
    }
```

problem here is that the object key ("name" in the person example) is not known here

**Reply by pedro-vicente** [reply]
This version of the program avoids reading the file, easier to run
```

#include <nlohmann/json.hpp>
#include <iostream>
#include <fstream>
using json = nlohmann::json;


int main()
{
  std::string str{ R"({ "Main": [ { "obj1": "bar" }, { "obj2": "foo" } ] })" };
  json js = json::parse(str);

  for (json::iterator it = js.begin(); it != js.end(); ++it)
  {
    std::cout << it.key() << " :\n";
    std::cout << it.value() << "\n";
  }

  if (js.contains("Main"))
  {
    json a = js["Main"];
    for (size_t idx = 0; idx < a.size(); ++idx)
    {
      json o = a.at(idx);
      std::cout << o << "\n";

    }
  }
  return 0;
}


```

**Reply by gregmarr** [reply]
You already have the code for iterating over the elements of a json array:

    json a = js["Main"];
    for (size_t idx = 0; idx < a.size(); ++idx)
    {
      json o = a.at(idx);
      std::cout << o << "\n";

    }

and the code for iterating over the keys and values of a json object:

    for (json::iterator it = js.begin(); it != js.end(); ++it)
    {
      std::cout << it.key() << " :\n";
      std::cout << it.value() << "\n";
    }

That gives you everything you need to create your key/value objects and add them to a vector.

**Reply by pedro-vicente** [reply]
Yes, but the API gives an automatic way to do this by means of NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE;
it seems it would be like just defining a vector like this, but this program gives the error

> 
[json.exception.out_of_range.403] key 'key' not found


program

```

#include <nlohmann/json.hpp>
#include <iostream>
#include <fstream>
using json = nlohmann::json;

struct key_value_t
{
  std::string key;
  std::string value;
};
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(key_value_t, key, value);

int main()
{
  try
  {
    std::string str{ R"({ "Main": [ { "obj1": "bar" }, { "obj2": "foo" } ] })" };
    json js = json::parse(str);

    std::vector<key_value_t> kv;
    if (js.contains("Main"))
    {
      js.at("Main").get_to(kv);
    }
  }

  catch (std::exception& e)
  {
    std::cout << e.what() << '\n';
  }
  return 0;
}
```

**Reply by gregmarr** [reply]
But that doesn't match your JSON structure.  That expects a JSON object that looks like `{ "key": "string 1", "value": "string 2" }`.  For this use case, you need to do it manually.

**Reply by gregmarr** [reply]
If you are sure that there is only a single key/value pair per object, you could potentially write your own `from_json` function like this.  Uncompiled code:

    void from_json(const json& j, key_value_t& kv) {
        assert(j.is_object());
        assert(j.size() == 1);
        kv.key = j.begin().key();
        kv.value = j.begin().value();
    }

**Reply by pedro-vicente** [reply]
@gregmarr that was exactly what I was looking for !, thanks

complete program, where I added a to_json() function to write output in the same format
```

#include <iostream>
#include <fstream>
#include <map>
#include <nlohmann/json.hpp>
#include <string>
#include <vector>
using json = nlohmann::json;

struct key_value_t
{
  std::string key;
  std::string value;
};

void from_json(const json& j, key_value_t& kv)
{
  assert(j.is_object());
  assert(j.size() == 1);
  kv.key = j.begin().key();
  kv.value = j.begin().value();
}

void to_json(json& j, const key_value_t& kv)
{
  j = nlohmann::json{
    {kv.key, kv.value}
  };
}

int main()
{
  try
  {
    std::string str{ R"({ "Main": [ { "obj1": "bar" }, { "obj2": "foo" } ] })" };
    json js_in = json::parse(str);

    std::vector<key_value_t> kv_in;

     //get input object
    js_in.at("Main").get_to(kv_in);

    if (js_in.contains("Main"))
    {
      nlohmann::json js_out = nlohmann::json::object_t();

      std::vector<key_value_t> kv_;
      key_value_t kv1{ "obj1", "1" };
      key_value_t kv2{ "obj2", "2" };
      kv_.push_back(kv1);
      kv_.push_back(kv2);

      nlohmann::json js_arr = kv_;

      //add to main JSON object
      js_out += {"Main", js_arr};

      //save output
      std::ofstream ofs("output.json");
      ofs << std::setw(2) << js_out << std::endl;
      ofs.close();

    }
  }

  catch (std::exception& e)
  {
    std::cout << e.what() << '\n';
  }
  return 0;
}
```


output file is
```

{
  "Main": [
    {
      "obj1": "1"
    },
    {
      "obj2": "2"
    }
  ]
}
```

**Reply by gregmarr** [reply]
Great.  Can you mark this question as answered?

**Reply by pedro-vicente** [reply]
@gregmarr How can I add your answer as "Mark as answer" ?

**Comment by gregmarr** [comment, marked as answer]
Apparently you can only mark top-level comments as answers.

If you are sure that there is only a single key/value pair per object, you could potentially write your own from_json function like this. Uncompiled code:

    void from_json(const json& j, key_value_t& kv) {
        assert(j.is_object());
        assert(j.size() == 1);
        kv.key = j.begin().key();
        kv.value = j.begin().value();
    }
